### ISSUE TITLE
EKS-A on Snow Slowness and performance Issues for Prod Grade Workloads.

### ISSUE THREAD

**Issue by elamaran11** [aws/eks-anywhere#5819]
**What happened**:

Set of issues i want to point out is on performance, slowness and consistent problems i get with my 1.24 EKS-A Snow cluster with 3 nodes CP and DP and im using this for my EKS-A Conformance framework to validate our ISV partners. Im using large instance size.

Issue #1, I consistently get the below issue while running Kubectl commands but sometimes the command work :

```
ubuntu@ip-34-223-14-199:~/eks-anywhere-conformance-testing$ kubectl get secret -n kube-system
Unable to connect to the server: dial tcp 192.168.4.75:6443: connect: no route to host
```

Issue #2, Im running some utlities like Kubecost, Vault etc etc and i get these for many pods :

```
  Warning  FailedScheduling  9m                 default-scheduler  0/6 nodes are available: 6 pod has unbound immediate PersistentVolumeClaims. preemption: 0/6 nodes are available: 6 Preemption is not helpful for scheduling.
```

Issue #3, Im getting the below on some pods and pod state is unavailable this happens when i restart my snow machine and instances :

```
  Warning  FailedCreatePodSandBox  117s                 kubelet  Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "a6ddc24c59c50407ccb605114901858b68828f79ac379cb1a9f69aa9fcf92c5a": plugin type="cilium-cni" name="cilium" failed (add): unable to connect to Cilium daemon: failed to create cilium agent client after 30.000000 seconds timeout: Get "http:///var/run/cilium/cilium.sock/v1/config": dial unix /var/run/cilium/cilium.sock: connect: no such file or directory
Is the agent running?
```

Issue #4, Whenever i see nodes i see the below. `InvalidDiskCapacity` alarms me :

```
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests      Limits
  --------           --------      ------
  cpu                1700m (85%)   4250m (212%)
  memory             2426Mi (30%)  5910Mi (75%)
  ephemeral-storage  2298Mi (10%)  4346Mi (19%)
  hugepages-1Gi      0 (0%)        0 (0%)
  hugepages-2Mi      0 (0%)        0 (0%)
Events:
  Type     Reason                   Age                From        Message
  ----     ------                   ----               ----        -------
  Normal   Starting                 10m                kube-proxy
  Warning  InvalidDiskCapacity      13m                kubelet     invalid capacity 0 on image filesystem
  Normal   Starting                 13m                kubelet     Starting kubelet.
  Normal   NodeAllocatableEnforced  13m                kubelet     Updated Node Allocatable limit across pods
  Normal   NodeHasNoDiskPressure    13m (x7 over 13m)  kubelet     Node eksa-conformance-snow-md-0-1-s62ph status is now: NodeHasNoDiskPressure
  Normal   NodeHasSufficientPID     13m (x7 over 13m)  kubelet     Node eksa-conformance-snow-md-0-1-s62ph status is now: NodeHasSufficientPID
  Normal   NodeHasSufficientMemory  13m (x8 over 13m)  kubelet     Node eksa-conformance-snow-md-0-1-s62ph status is now: NodeHasSufficientMemory
```

Issue #5, My PVCs sometimes goes to permanent pending state.

Issue #6, Creation of ClusterSecretStore fails, please see below :

```
Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "validate.clustersecretstore.external-secrets.io": failed to call webhook: Post "[https://external-secrets-webhook.external-secrets.svc:443/validate-external-secrets-io-v1beta1-clustersecretstore?timeout=5s](https://external-secrets-webhook.external-secrets.svc/validate-external-secrets-io-v1beta1-clustersecretstore?timeout=5s)": dial tcp 10.101.94.229:443: connect: connection refused
Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "validate.clustersecretstore.external-secrets.io": failed to call webhook: Post "[https://external-secrets-webhook.external-secrets.svc:443/validate-external-secrets-io-v1beta1-clustersecretstore?timeout=5s](https://external-secrets-webhook.external-secrets.svc/validate-external-secrets-io-v1beta1-clustersecretstore?timeout=5s)": dial tcp 10.101.94.229:443: connect: connection refused
clusterexternalsecret.external-secrets.io/aws-env-metadata created
```

Im trying to run our conformance harness to validate partner addons on EKS-A on Snow cluster. So we are blocked in validating our partners time to time. 

We want to know what is the best recommendation configuration in terms of running our Conformance harness at scale like how we run Prod workloads :

- How many snow machines are required?
- What is the recommend Instance size, number of nodes for CP and DP.
- Currently if we follow the EKS-D image creation process, it is 50 GB image which we think is one of the reasons for slowness. Is there a way to create a slimmed down version of image
- What should be the Cluster configuration, K8s version, node instance types etc etc to avoid any of the above issues
- What should we do to avoid networking issues that we see above
- What should we do to avoid storage issues etc etc
- When do we have snapshotting support in EBS

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- EKS Anywhere Release:
- EKS Distro Release:

**Comment by csplinter** [comment]
@elamaran11 please keep the environment running. Will be helpful to know the steps you performed that resulted in these issues and any error logs if you've dug into them. Lots of different things in this ticket and we need to go thru 1 by 1.

**Comment by elamaran11** [comment]
@csplinter Sounds good. I have environment up and running. Im following the standard process shared to create and side load images. Then eventually running our Conformance framework. Happy to jump on a session and share the environment issues i have.

**Comment by elamaran11** [comment]
Hi @jiayiwang7 @yxinchen Based on your recommendations yesterday, the cluster is a lot faster now compared to previous versions. But i still see a lot of interim network failures which i think we need to do something about. Please see some samples below :

```
ubuntu@ip-34-223-14-199:~$ kubectl replace --raw "/api/v1/namespaces/kubecost/finalize" -f ./kubecost.json
Unable to connect to the server: dial tcp 192.168.4.75:6443: connect: no route to host
```

```
ubuntu@ip-34-223-14-199:~$ k get ns
Unable to connect to the server: dial tcp 192.168.4.75:6443: connect: no route to host
ubuntu@ip-34-223-14-199:~$ k get ns
NAME                                  STATUS        AGE
...
```


```
ubuntu@ip-34-223-14-199:~$ kubectl get namespace vault -o json > vault.json
Unable to connect to the server: dial tcp 192.168.4.75:6443: connect: no route to host
ubuntu@ip-34-223-14-199:~$ kubectl get namespace vault -o json > vault.json
Unable to connect to the server: dial tcp 192.168.4.75:6443: connect: no route to host
ubuntu@ip-34-223-14-199:~$ kubectl get namespace vault -o json > vault.json
ubuntu@ip-34-223-14-199:~$ micro vault.json
ubuntu@ip-34-223-14-199:~$ kubectl replace --raw "/api/v1/namespaces/vault/finalize" -f ./vault.json
{"kind":"Namespace","apiVersion":"v1","metadata":{"name":"vault","uid":"56c51851-d9e2-48e2-874e-97d18290b7a8"
```
After running my Conformance, the cluster has again gone back to slowness but not like before. And whenever i have workload with storage using EBS SC, the pod takes lot of time to come back. Please see below 10 mins and still it isnt `bound` :

```ubuntu@ip-34-223-14-199:~$ k get pvc -A
NAMESPACE   NAME                                  STATUS    VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
kubecost    kubecost-kubecost-cost-analyzer       Bound     pvc-eded0a14-d6c2-4c8e-8617-2f2e64d7406b   32Gi       RWO            ebs-sc         10m
kubecost    kubecost-kubecost-prometheus-server   Pending   pvc-29788935-ecc9-41ee-b6e5-6531df239258   0                         ebs-sc         10m
vault       data-vault-vault-0                    Bound     pvc-eb0e3ace-1a6d-4b10-b745-da15a9b3c7ad   10Gi       RWO            ebs-sc         15m
```
I would appreciate if we can make something to get this running faster. This is as same as Prod workload.

**Comment by elamaran11** [comment]
Also do  have a newer approach to EBS storage on Snow?

**Comment by elamaran11** [comment]
Based on our chat, im adding more info on Cilium Operator failure with etcd. Im following the setup tools to create the cluster. If stacked etcd is not right approach as per the setup tools for snow, appreciate if you can provide some guidance so i can fix this issue either with upgrade or via new cluster.

```ubuntu@ip-34-223-14-199:~$ k logs cilium-operator-6c8bfc95dd-xs8hw -n kube-system
level=info msg="Started gops server" address="127.0.0.1:9891" subsys=cilium-operator-generic
level=info msg="  --cmdref=''" subsys=cilium-operator-generic
level=info msg="  --config=''" subsys=cilium-operator-generic
level=info msg="  --config-dir='/tmp/cilium/config-map'" subsys=cilium-operator-generic
level=info msg="  --debug='false'" subsys=cilium-operator-generic
level=info msg="  --log-opt=''" subsys=cilium-operator-generic
level=info msg="  --version='false'" subsys=cilium-operator-generic
level=info msg="Cilium Operator 1.11.10-eksa.2 a81aec9 2022-10-18T16:04:48+00:00 go version go1.17.13 linux/amd64" subsys=cilium-operator-generic
level=info msg="Registering Operator metrics" subsys=metrics
level=info msg="Starting apiserver on address 127.0.0.1:9234" subsys=cilium-operator-api
level=info msg="Establishing connection to apiserver" host="[https://10.96.0.1:443](https://10.96.0.1/)" subsys=k8s
level=info msg="Connected to apiserver" subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumEndpoint/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumIdentity/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumNode/v2 subsys=k8s
level=info msg="attempting to acquire leader lease default/cilium-operator-resource-lock..." subsys=klog
level=info msg="Leader re-election complete" newLeader=eksa-conformance-snow-control-plane-1-zx8js-VARDfxBfSn operatorID=eksa-conformance-snow-control-plane-1-8sx7j-RQsiKrcITT subsys=cilium-operator-generic
level=info msg="successfully acquired lease default/cilium-operator-resource-lock" subsys=klog
level=info msg="Leading the operator HA deployment" subsys=cilium-operator-generic
level=info msg="Initializing IPAM" mode=kubernetes subsys=cilium-operator-generic
level=info msg="Removing Cilium Node Taints or Setting Cilium Is Up Condition for Kubernetes Nodes" k8sNamespace=default label-selector="k8s-app=cilium" remove-cilium-node-taints=true set-cilium-is-up-condition=true subsys=cilium-operator-generic
level=info msg="Starting to synchronize CiliumNode custom resources" subsys=cilium-operator-generic
level=info msg="Starting to garbage collect stale CiliumNode custom resources" subsys=watchers
level=info msg="Starting to garbage collect stale CiliumEndpoint custom resources" subsys=cilium-operator-generic
level=info msg="Starting CRD identity garbage collector" interval=15m0s subsys=cilium-operator-generic
level=info msg="CiliumNodes caches synced with Kubernetes" subsys=cilium-operator-generic
level=info msg="Initialization complete" subsys=cilium-operator-generic
level=warning msg="Network status error received, restarting client connections" error="an error on the server (\"[+]ping ok\\n[+]log ok\\n[-]etcd failed: reason withheld\\n[+]poststarthook/start-kube-apiserver-admission-initializer ok\\n[+]poststarthook/generic-apiserver-start-informers ok\\n[+]poststarthook/priority-and-fairness-config-consumer ok\\n[+]poststarthook/priority-and-fairness-filter ok\\n[+]poststarthook/start-apiextensions-informers ok\\n[+]poststarthook/start-apiextensions-controllers ok\\n[+]poststarthook/crd-informer-synced ok\\n[+]poststarthook/bootstrap-controller ok\\n[+]poststarthook/rbac/bootstrap-roles ok\\n[+]poststarthook/scheduling/bootstrap-system-priority-classes ok\\n[+]poststarthook/priority-and-fairness-config-producer ok\\n[+]poststarthook/start-cluster-authentication-info-controller ok\\n[+]poststarthook/aggregator-reload-proxy-client-cert ok\\n[+]poststarthook/start-kube-aggregator-informers ok\\n[+]poststarthook/apiservice-registration-controller ok\\n[+]poststarthook/apiservice-status-available-controller ok\\n[+]poststarthook/kube-apiserver-autoregistration ok\\n[+]autoregister-completion ok\\n[+]poststarthook/apiservice-openapi-controller ok\\n[+]poststarthook/apiservice-openapiv3-controller ok\\nhealthz check failed\") has prevented the request from succeeding (get healthz.meta.k8s.io)" subsys=k8s
level=warning msg="Network status error received, restarting client connections" error="an error on the server (\"[+]ping ok\\n[+]log ok\\n[-]etcd failed: reason withheld\\n[+]poststarthook/start-kube-apiserver-admission-initializer ok\\n[+]poststarthook/generic-apiserver-start-informers ok\\n[+]poststarthook/priority-and-fairness-config-consumer ok\\n[+]poststarthook/priority-and-fairness-filter ok\\n[+]poststarthook/start-apiextensions-informers ok\\n[+]poststarthook/start-apiextensions-controllers ok\\n[+]poststarthook/crd-informer-synced ok\\n[+]poststarthook/bootstrap-controller ok\\n[+]poststarthook/rbac/bootstrap-roles ok\\n[+]poststarthook/scheduling/bootstrap-system-priority-classes ok\\n[+]poststarthook/priority-and-fairness-config-producer ok\\n[+]poststarthook/start-cluster-authentication-info-controller ok\\n[+]poststarthook/aggregator-reload-proxy-client-cert ok\\n[+]poststarthook/start-kube-aggregator-informers ok\\n[+]poststarthook/apiservice-registration-controller ok\\n[+]poststarthook/apiservice-status-available-controller ok\\n[+]poststarthook/kube-apiserver-autoregistration ok\\n[+]autoregister-completion ok\\n[+]poststarthook/apiservice-openapi-controller ok\\n[+]poststarthook/apiservice-openapiv3-controller ok\\nhealthz check failed\") has prevented the request from succeeding (get healthz.meta.k8s.io)" subsys=k8s
Failed to update lock: Put "[https://10.96.0.1:443/apis/coordination.k8s.io/v1/namespaces/default/leases/cilium-operator-resource-lock](https://10.96.0.1/apis/coordination.k8s.io/v1/namespaces/default/leases/cilium-operator-resource-lock)": context deadline exceeded
level=error msg="Failed to update lock: Put \"[https://10.96.0.1:443/apis/coordination.k8s.io/v1/namespaces/default/leases/cilium-operator-resource-lock](https://10.96.0.1/apis/coordination.k8s.io/v1/namespaces/default/leases/cilium-operator-resource-lock)\": context deadline exceeded" subsys=klog
level=info msg="failed to renew lease default/cilium-operator-resource-lock: timed out waiting for the condition" subsys=klog
level=info msg="Leader election lost" operator-id=eksa-conformance-snow-control-plane-1-8sx7j-RQsiKrcITT subsys=cilium-operator-generic
```

**Comment by elamaran11** [comment]
Thankyou @jiayiwang7 @yxinchen for troubleshooting this issue with me last week. As we tried different combos like 
- Using 2xl instances for Data Plane and xl instances for Control Plane
- Adding Container Volume of 500 GB for Data Plane and 100 GB for Control Plane
- UnStacked ETCD
- Unstacked ETCD with Static IP for all machines
- Unstacked ETCD with static Ip for ETCD and DHCP for Control/Data Planes

We got a stable cluster working and conformance framework is working on EKS-A on Snowball. Thankyou so much. I will keep you posted as i hit on any performance issues.
